lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -3159,6 +3480,96 @@ SQLite3_result * MySQL_HostGroups_Manager::SQL3_Free_Connections() {
 	return result;
 }
 
+void MySQL_HostGroups_Manager::p_update_connection_pool_update_counter(std::string& endpoint_id, std::string& endpoint_addr, std::string& endpoint_port, std::string& hostgroup_id, std::map<std::string, prometheus::Counter*>& m_map, unsigned long long value, p_hg_dyn_counter::metric idx) {
+	const auto& counter_id = m_map.find(endpoint_id);
+	if (counter_id != m_map.end()) {
+		const auto& cur_val = counter_id->second->Value();
+		counter_id->second->Increment(value - cur_val);
+	} else {
+		auto& new_counter = status.p_dyn_counter_array[idx];
+		m_map.insert(
+			{
+				endpoint_id,
+				std::addressof(new_counter->Add({
+					{"endpoint", endpoint_addr + ":" + endpoint_port},
+					{"hostgroup", hostgroup_id }
+				}))
+			}
+		);
+	}
+}
+
+void MySQL_HostGroups_Manager::p_update_connection_pool_update_gauge(std::string& endpoint_id, std::string& endpoint_addr, std::string& endpoint_port, std::string& hostgroup_id, std::map<std::string, prometheus::Gauge*>& m_map, unsigned long long value, p_hg_dyn_gauge::metric idx) {
+	const auto& counter_id = m_map.find(endpoint_id);
+	if (counter_id != m_map.end()) {
+		counter_id->second->Set(value);
+	} else {
+		auto& new_counter = status.p_dyn_gauge_array[idx];
+		m_map.insert(
+			{
+				endpoint_id,
+				std::addressof(new_counter->Add({
+					{"endpoint", endpoint_addr + ":" + endpoint_port},
+					{"hostgroup", hostgroup_id }
+				}))
+			}
+		);
+	}
+}
+
+void MySQL_HostGroups_Manager::p_update_connection_pool() {
+	wrlock();
+	for (int i = 0; i < static_cast<int>(MyHostGroups->len); i++) {
+		MyHGC *myhgc = static_cast<MyHGC*>(MyHostGroups->index(i));
+		for (int j = 0; j < static_cast<int>(myhgc->mysrvs->cnt()); j++) {
+			MySrvC *mysrvc = static_cast<MySrvC*>(myhgc->mysrvs->servers->index(j));
+			std::string endpoint_addr = mysrvc->address;
+			std::string endpoint_port = std::to_string(mysrvc->port);
+			std::string hostgroup_id = std::to_string(myhgc->hid);
+			std::string endpoint_id = hostgroup_id + ":" + endpoint_addr + ":" + endpoint_port;
+
+			// proxysql_connection_pool_bytes_data_recv metric
+			p_update_connection_pool_update_counter(endpoint_id, endpoint_addr, endpoint_port, hostgroup_id,
```

Comment:
`p_update_connection_pool()` calls `p_update_connection_pool_update_counter()` and `p_update_connection_pool_update_gauge()` .
The last 2 are responsible for finding the endpoint in the map, and if the endpoint doesn't exist , it creates it.

What about if the server is removed? (it doesn't exist any longer).
I think we should stop sending metrics from a server that doesn't exist any longer.

Do we have a way to de-register a counter?

This comments applies also to gtid_executed table